### PR TITLE
Remove pod/node scheduling correctly

### PR DESCRIPTION
**Pull request by DaiYuzeng**
### Summary
- https://github.com/rancher/dashboard/issues/6366

### CODE REVIEW

**Comment by WuJun2016** [comment]
@mantis-toboggan-md    PTAL thanks.

File: shell/components/form/NodeAffinity.vue [hunk 1]
```diff
@@ -50,7 +51,10 @@ export default {
       return {
         allSelectorTerms,
         weightedNodeSelectorTerms: preferredDuringSchedulingIgnoredDuringExecution,
-        defaultWeight:             1
+        defaultWeight:             1,
+        // rules in MatchExpressions.vue can not catch changes what happens on parent component
+        // we need re-render it via key changing
+        rerenderNums:              0
```
Comment (aalves08) [inline shell/components/form/NodeAffinity.vue:57]: Just tested your code and this is not needed... The secret to your fix lies on the `@remove` event at the `ArrayListGrouped` component. All references to `rerenderNums` should be deleted from the PR

File: shell/components/form/NodeAffinity.vue [hunk 2]
```diff
@@ -73,35 +77,38 @@ export default {
...
+        out.preferredDuringSchedulingIgnoredDuringExecution = preferredDuringSchedulingIgnoredDuringExecution;
+      }
+      if (requiredDuringSchedulingIgnoredDuringExecution.nodeSelectorTerms.length) {
+        out.requiredDuringSchedulingIgnoredDuringExecution = requiredDuringSchedulingIgnoredDuringExecution;
+      }
+      this.$emit('input', out);
+    },
+
+    remove() {
+      this.rerenderNums += 1;
```
Comment (aalves08) [inline shell/components/form/NodeAffinity.vue:110]: to delete

File: shell/components/form/NodeAffinity.vue [hunk 3]
```diff
@@ -142,6 +156,7 @@ export default {
             </div>
           </div>
           <MatchExpressions
+            :key="rerenderNums"
```
Comment (aalves08) [inline shell/components/form/NodeAffinity.vue:159]: to delete

File: shell/components/form/PodAffinity.vue [hunk 4]
```diff
@@ -71,7 +71,10 @@ export default {
 
     return {
       allSelectorTerms,
-      defaultWeight: 1
+      defaultWeight: 1,
+      // rules in MatchExpressions.vue can not catch changes what happens on parent component
+      // we need re-render it via key changing
+      rerenderNums:  0
```
Comment (aalves08) [inline shell/components/form/PodAffinity.vue:77]: to delete

File: shell/components/form/PodAffinity.vue [hunk 5]
```diff
@@ -134,6 +138,11 @@ export default {
       Object.assign(this.value.affinity, { podAffinity, podAntiAffinity });
     },
 
+    remove() {
+      this.rerenderNums += 1;
```
Comment (aalves08) [inline shell/components/form/PodAffinity.vue:142]: to delete

File: shell/components/form/PodAffinity.vue [hunk 6]
```diff
@@ -235,6 +245,7 @@ export default {
             />
           </div>
           <MatchExpressions
+            :key="rerenderNums"
```
Comment (aalves08) [inline shell/components/form/PodAffinity.vue:248]: to delete

**Review by aalves08 — COMMENTED** [review]
apart from the `rerenderNums` that can be deleted, otherwise code LGTM. Thanks

**Comment by WuJun2016** [comment]
Hi， @aalves08  rerenderNums works to solve this problem.


https://user-images.githubusercontent.com/24985926/179504550-affae706-8a0a-4a46-b5f8-6888ee6552ef.mov

**Comment by DaiYuzeng** [comment]
> > Hi， @aalves08 rerenderNums works to solve this problem.
> > fix.mov
> 
> That's not the desired behaviour... If you delete the first block, you should only have left the second block (key: 2, val: 2).
> 
> I tried again your code without the `rerenderNums` and everything seems to work as expected.
> 
> FYI @nwmac @richard-cox

@aalves08 @WuJun2016 Hi! It seems like the issue of remove still occured if removed `rerenderNums`, reproduced on master branch of rancher/dashboard

Kapture 2022-07-19 at 09 52 24

**Comment by aalves08** [comment]
Ok, there seems to be a misunderstanding from my side (although I tested this but probably didn't refresh it properly or saved... 🤷 ).
Your key thing is 100% correct. I tried without it and although the payload sent is correct, it didn't update the UI properly because it doesn't have any reference (key) to it and it loses it's bearing.
I had to investigate why, because this is a strange pattern to be seen, but it turns out that we have a `v-for` on `ArrayListGrouped` component, for which the key can only be passed to it's children, such as `MatchExpressions` in this case, because we pass them via scoped slots.

Without it, the UI just loses it's context and renders the first item that it had stored on the dom... strange but true, although the payload is kept intact.

So, I retract my earlier comments 🙏 and my final comment would be to create the key via a method ex: `generateKey` that generates a random string ex: `return randomStr(15).toLowerCase();`. be aware that the function must be called on both the `data` variable and the `remove` function.

I would prefer to see this change, because we will never know where we are going to apply this component again and we don't want duplicate key issues.

Otherwise, code LGTM. Sorry and thank you.

Tagging both @nwmac and @richard-cox for FYI and if they want to weigh in.

**Comment by mantis-toboggan-md** [comment]
Can we close this in favor of https://github.com/rancher/dashboard/pull/6398?

**Comment by DaiYuzeng** [comment]
Close this PR at first. We can find out this commit on https://github.com/rancher/dashboard/pull/6398, thanks!

**Pull request closed without merging**